lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -1519,41 +1519,37 @@ void ConstantOp::build(OpBuilder &builder, OperationState &result,
   return build(builder, result, type, attr);
 }
 
+static LogicalResult verifySubfieldOp(SubfieldOp op) {
+  if (op.fieldIndex() >=
+      op.input().getType().cast<BundleType>().getNumElements())
+    return op.emitOpError("subfield element index is greater than the number "
+                          "of fields in the bundle type");
+  return success();
+}
+
 FIRRTLType SubfieldOp::inferReturnType(ValueRange operands,
                                        ArrayRef<NamedAttribute> attrs,
                                        Optional<Location> loc) {
-  auto inType = operands[0].getType();
-  auto fieldname = getAttr<StringAttr>(attrs, "fieldname");
+  auto inType = operands[0].getType().cast<BundleType>();
+  auto fieldIndex =
+      getAttr<IntegerAttr>(attrs, "fieldIndex").getValue().getZExtValue();
 
-  if (auto bundleType = inType.dyn_cast<BundleType>()) {
-    auto elt = bundleType.getElement(fieldname.getValue());
-    if (!elt) {
-      if (loc)
-        mlir::emitError(*loc, "unknown field '")
-            << fieldname.getValue() << "' in bundle type " << inType;
-      return {};
-    }
-    // FIRRTL puts flips on element fields, not on the underlying
-    // types.  The result type of a subfield should strip a flip
-    // if one exists.
-    return elt->type;
+  if (fieldIndex >= inType.getNumElements()) {
+    if (loc)
+      mlir::emitError(*loc, "subfield element index is greater than the number "
+                            "of fields in the bundle type");
+    return {};
   }
 
-  if (loc)
-    mlir::emitError(*loc, "subfield requires bundle operand");
-  return {};
+  // SubfieldOp verifier checks that the field index is valid with number of
+  // subelements.
+  return inType.getElement(fieldIndex).type;
 }
 
 bool SubfieldOp::isFieldFlipped() {
-  auto fieldname = this->fieldname();
+  auto fieldIndex = this->fieldIndex();
```

Comment:
Nice.  I think it would make sense to inline fieldIndex now and drop the `this->`